### PR TITLE
Dataflow: Add path-problem view of intermediate stages for debug purposes.

### PR DESCRIPTION
**Pull request by aschackmull**
This adds a module `Debug` available in a data flow module resulting from the application of `Global<..>` to a configuration. The `Debug` module contains references to the individual stages, which in turn contain a `Graph` module with a `query edges` predicate and a `query subpaths` predicate. The associated PathNode type also contains convenience predicate for picking out sources and sinks, and additionally arbitrary-source and arbitrary-sink in case there's an abundance of flow and one just wants to see some representative examples.

### CODE REVIEW

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -2487,6 +2486,264 @@ module MakeImpl<LocationSig Location, InputSig<Location> Lang> {
           callEdgeReturn(call, c, _, _, _, _, _)
         }
 
+        additional module Graph {
```
Comment (hvitved) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:2489]: Perhaps add QL doc that this is for debugging only.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -2487,6 +2486,264 @@ module MakeImpl<LocationSig Location, InputSig<Location> Lang> {
...
+              NodeEx node, FlowState state, Cc cc, ParamNodeOption summaryCtx, TypOption argT,
+              ApOption argAp, Typ t, Ap ap
+            ) {
+              fwdFlow(node, state, cc, summaryCtx, argT, argAp, t, ap, _) and
+              revFlow(node, state, _, _, ap)
+            } or
+            TStagePathNodeSrcGrp() or
+            TStagePathNodeSinkGrp()
+
+          class StagePathNode extends TStagePathNode {
```
Comment (hvitved) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:2501]: I suspect that we will want to have `Node getNode()` as well.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 3]
```diff
@@ -4639,67 +4896,127 @@ module MakeImpl<LocationSig Location, InputSig<Location> Lang> {
...
+        tftuples = -1
+        or
+        stage = "1 Rev" and
+        n = 15 and
+        Stage1::stats(false, nodes, fields, conscand, states, tuples, calledges) and
+        tfnodes = -1 and
+        tftuples = -1
+      }
+
+      predicate stageStats2(
```
Comment (hvitved) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:4940]: Yes, I have been manually editing the `stageStats` predicate to cap the stages; good to have this builtin.

**Review by hvitved — APPROVED** [review]
LGTM; one suggestion. Also, given the change to `fwdFlowIsEntered`, I think we should run DCA for all languages.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 1 above)
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:2489]: It's only exposed through the `Debug` module. But a guess an additional piece of qldoc here improves local readability. I'll add it.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:2501]: Yes, I'll add it.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 3 above)
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:4940]: So have I 😄

**Comment by aschackmull** [comment]
dca looks uneventful as expected.

**Review by hvitved — APPROVED** [review]

**Pull request merged**